src/sessions/p2p_session.rs

```diff
@@ -192,6 +192,15 @@ impl<T: Config> P2PSession<T> {
             SessionState::Synchronizing
         };
 
+        let sparse_saving = if max_prediction == 0 {
+            // in lockstep mode, saving will never happen, but we use the last saved frame to mark
+            // control marking frames confirmed, so we need to turn off sparse saving to ensure that
+            // frames are marked as confirmed - otherwise we will never advance the game state.
+            false
```

Comment:
Ideally we'd log a warning (or at least at debug level) here to let the user know their sparse saving request is being ignored.